### PR TITLE
mcid.py: optimize FFT and A-weighting calculations

### PR DESCRIPTION
**Pull request by deanlee**
Precompute A-weighting filter coefficients for reuse in each callback to enhance performance and reduce computational overhead.

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
Do you have before/after on some benchmark? We need that for any optimization PR.

**Comment by deanlee** [comment]
I have not tested the reduced CPU usage of this PR on the device separately. I only measured the reduced time, and it saves approximately 1.5 ms on every update on the device.

As for CPU usage, I tested it along with another PR (not yet submitted) that replaces np.concatenate with a circular buffer to efficiently manage incoming audio samples, avoiding costly array resizing operations. Together, these two PRs can reduce CPU usage by half (about reduce 4% cpu usage)

**Comment by adeebshihadeh** [comment]
`test_onroad.py` is a good benchmark for this https://github.com/commaai/openpilot/blob/0fa6745a67b60a572cfa4060d8155ac50f4e4e06/selfdrive/test/test_onroad.py#L59

**Comment by adeebshihadeh** [comment]
trigger-jenkins

**Comment by adeebshihadeh** [comment]
test_onroad says the CPU went to 3.9% from 5%, so definitely an improvement! But the memory usage check failed:
image

**Comment by deanlee** [comment]
Not sure what the issue is. The cached calculation result A_WEIGHTING in the global variable is only about 4KB.

File: system/micd.py [hunk 1]
```diff
@@ -12,6 +12,15 @@
...
 
+def precompute_a_weighting():
+  # Calculate the A-weighting filter
+  # https://en.wikipedia.org/wiki/A-weighting
+  freqs = np.fft.fftfreq(FFT_SAMPLES, d=1 / SAMPLE_RATE)
+  A = 12194 ** 2 * freqs ** 4 / ((freqs ** 2 + 20.6 ** 2) * (freqs ** 2 + 12194 ** 2) * np.sqrt((freqs ** 2 + 107.7 ** 2) * (freqs ** 2 + 737.9 ** 2)))
+  return A / np.max(A)
+
+
+A_WEIGHTING = precompute_a_weighting()
```
Comment (sshane) [inline system/micd.py:23]: is caching not better in your opinion?

Comment (deanlee) [reply system/micd.py:23]: The filter is computed once and reused, so a global variable works well. 
Using a global variable for caching is likely the better approach here for the following reason:
1. The result doesn’t change with input, so no need for dynamic caching.
2. Global variables are simple,straightforward and avoid the overhead of cache decorators when results are fixed.

Comment (deanlee) [reply system/micd.py:23]: > caching seems cleaner to me

I agree, caching does offer a cleaner approach. If you think it's the way to go, let's use that.

**Comment by sshane** [comment]
jenkins passed on a subsequent run @adeebshihadeh

**Comment by adeebshihadeh** [comment]
I haven't seen that test fail randomly before. We can run a few more times to confirm though

**Comment by sshane** [comment]
Passed 5 runs across the day. @deanlee can you make it use cache?

**Comment by deanlee** [comment]
Completed. The function now uses cache.

**Pull request merged**